crypto/s2n_pkey.c

```diff
@@ -104,26 +106,32 @@ int s2n_pkey_match(const struct s2n_pkey *pub_key, const struct s2n_pkey *priv_k
     return pub_key->match(pub_key, priv_key);
 }
 
-int s2n_pkey_free(struct s2n_pkey *pkey)
+int s2n_pkey_free(struct s2n_pkey *key)
 {
-    if (pkey == NULL || pkey->free == NULL) {
-        return 0;
+    if (key != NULL && key->free != NULL) {
+        GUARD(key->free(key));
```

Comment:
Does `key->free(key)` potentially invalidate memory sufficiently that calling `EVP_PKEY_free(key->pkey)` later is dangerous?

I'm okay trusting that `key->free(key)` is sufficient. I'm okay explicitly calling `EVP_KPEY_free(key->pkey)` and _then_  `key->free(key)`,  but this order seems dangerous.

Reply:
key->free(key) should never be sufficient. key->pkey is managed by s2n_pkey, not by any of the specific implementations. key->free(key) is for them to free any legacy / implementation specific stuff (aka, key->rsa_key and key->ec_key) they allocated -- that's why the rsa-pss free does nothing, because it allocates nothing. pkey belongs to the generic implementation, so shouldn't be freed by the specific implementations.

I'm not sure how either order is more or less dangerous.